nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,172 @@
+{
+
+PolygonSource::PolygonSource(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string & source_name,
+  const std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+  const std::string & base_frame_id,
+  const std::string & global_frame_id,
+  const tf2::Duration & transform_tolerance,
+  const rclcpp::Duration & source_timeout,
+  const bool base_shift_correction)
+: Source(
+    node, source_name, tf_buffer, base_frame_id, global_frame_id,
+    transform_tolerance, source_timeout, base_shift_correction),
+  data_(nullptr)
+{
+}
+
+PolygonSource::~PolygonSource()
+{
+  data_sub_.reset();
+}
+
+void PolygonSource::configure()
+{
+  Source::configure();
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string source_topic;
+
+  getParameters(source_topic);
+
+  rclcpp::QoS qos = rclcpp::SensorDataQoS();  // set to default
+  data_sub_ = node->create_subscription<nav2_msgs::msg::PolygonsArray>(
+    source_topic, qos,
+    std::bind(&PolygonSource::dataCallback, this, std::placeholders::_1));
+}
+
+void PolygonSource::getData(
+  const rclcpp::Time & curr_time,
+  std::vector<Point> & data) const
+{
+  // Ignore data from the source if it is not being published yet or
+  // not published for a long time
+  if (data_ == nullptr || data_->polygons.empty()) {
+    return;
+  }
+  // get the oldest time stamp from the polygon array
+  rclcpp::Time oldest_stamp = rclcpp::Time(data_->polygons[0].header.stamp);
+  for (const auto & polygon : data_->polygons) {
+    if (rclcpp::Time(polygon.header.stamp) < oldest_stamp) {
+      oldest_stamp = rclcpp::Time(polygon.header.stamp);
+    }
+  }
+  if (!sourceValid(oldest_stamp, curr_time)) {
+    return;
+  }
+
+  tf2::Stamped<tf2::Transform> tf_transform;
+
+  for (const auto & polygon : data_->polygons) {
+    if (base_shift_correction_) {
+      // Obtaining the transform to get data from source frame and time where it was received
+      // to the base frame and current time
+      if (
+        !nav2_util::getTransform(
+          polygon.header.frame_id, polygon.header.stamp,
+          base_frame_id_, curr_time, global_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return;
+      }
+    } else {
+      // Obtaining the transform to get data from source frame to base frame without time shift
+      // considered. Less accurate but much more faster option not dependent on state estimation
+      // frames.
+      if (
+        !nav2_util::getTransform(
+          polygon.header.frame_id, base_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return;
+      }
+    }
+    geometry_msgs::msg::PolygonStamped poly_out;
+    geometry_msgs::msg::TransformStamped tf = tf2::toMsg(tf_transform);
+    tf2::doTransform(polygon, poly_out, tf);
+    convertPolygonStampedToPoints(poly_out, data);
+  }
+
+}
+void PolygonSource::convertPolygonStampedToPoints(
```

Comment:
Space between functions

Overall, linting / styling here is weird. Please fix.

Reply:
I ran ament_uncrustify on it, please specify what you consider weird so I can fix it.